### PR TITLE
Basic support for the Haiku operating system

### PR DESCRIPTION
**Pull request by pulkomandy**
https://bugs.webkit.org/show_bug.cgi?id=220690

Reviewed by NOBODY (OOPS!).

Trivial changes to add Haiku as a known operating system in places where it uses already existing code in WTF. This will make maintenance of the Haiku port of WebKit easier by avoiding conflicts whenever another OS is added or removed. It will also benefit other ports of WebKit (for example GTK WebKit) when compiled on Haiku, as they also need these changes.

* Source/WTF/wtf/InlineASM.h:
* Source/WTF/wtf/MemoryPressureHandler.cpp: (WTF::MemoryPressureHandler::MemoryPressureHandler):
* Source/WTF/wtf/MemoryPressureHandler.h:
* Source/WTF/wtf/NumberOfCores.cpp: (WTF::numberOfProcessorCores):
* Source/WTF/wtf/PlatformEnable.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformLegacy.h:
* Source/WTF/wtf/PlatformOS.h:
* Source/WTF/wtf/RawHex.h:
* Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp: (WTF::OSAllocator::tryReserveUncommitted):
(WTF::OSAllocator::commit):
(WTF::OSAllocator::decommit):
https://github.com/WebKit/WebKit/commit/97b667dec2032f21192ef5bfc724600c1189571e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66038 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45411 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18657 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70069 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16647 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53210 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16929 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53007 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11590 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41899 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57170 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33642 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38570 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14547 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15524 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/59152 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60456 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14894 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71772 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/65282 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9993 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14307 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60322 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10025 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57232 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60612 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8253 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1892 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/87049 "Built successfully") | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10168 "") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41219 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15319 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42295 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43478 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42039 "Built successfully") | | |

### CODE REVIEW

**Comment by pulkomandy** [comment]
@aperezdc hello, as discussed previously, I'll (finally!) try to upstream the Haiku port again. I will start with simple changes that should hopefully not cause too much problems. I will make larger changes later, starting with WTF to add a few other things that need new Haiku-specific code (getting current time, renaming threads, etc).

File: Source/WTF/wtf/PlatformHave.h [hunk 1]
```diff
@@ -429,7 +429,7 @@
 #define HAVE_FOUNDATION_WITH_SAME_SITE_COOKIE_SUPPORT 1
 #endif
 
-#if PLATFORM(COCOA) || PLATFORM(GTK) || PLATFORM(WPE)
+#if PLATFORM(COCOA) || PLATFORM(GTK) || PLATFORM(HAIKU) || PLATFORM(WPE)
```
Comment (aperezdc) [inline Source/WTF/wtf/PlatformHave.h:432]: Nice, I didn't know Haiku had dark mode support these days! (I should try out a recent build some day, last time was about a year ago 😇 )

**Review by aperezdc — APPROVED** [review]
This looks reasonable and low-friction to me. I would expect this is enough to build JSCOnly on Haiku, right? 

Also, in case anybody else checking the PR wonders: yes, there has been builds of [WebKitGTK](https://depot.haiku-os.org/#!/pkg/webkit_gtk/haikuports/haikuports_x86_64/2/40/0/-/3/x86_64) + [GNOME Web](https://depot.haiku-os.org/#!/pkg/epiphany/haikuports/haikuports_x86_64/43/1/-/-/1/x86_64) for Haiku available for a while.

File: Source/WTF/wtf/PlatformHave.h (hunk 1 above)
Comment (pulkomandy) [reply Source/WTF/wtf/PlatformHave.h:432]: Yes, if you set your system colors to be dark (this is much easier to do in upcoming Haiku beta 5, we have simplified the Appearance preferences), WebKit will detect it and use dark mode when available

**Comment by pulkomandy** [comment]
I think a few more changes will be needed to build the jsconly port: 

- Some parts of WTF require Haiku specific code (for example, to get the system time, we have better/simpler APIs than what's used on Linux), I will submit those separately (or I can update this merge request with those if that's preferred)
- Maybe some changes in JavascriptCore (I have to check that)
- Some changes are needed in the build scripts (Tools/Scripts) to adjust a few paths that are different (but I have some work to check some old patchess that may not be needed anymore)

**Comment by aperezdc** [comment]
@pulkomandy haven't added the `request-merge-queue` label, but I suppose you would like me to add this to the commit queue, isn't it?

**Comment by pulkomandy** [comment]
Yes, I think this part is ready to merge :)
Is there something I need to do to "formally" request that?

**Comment by aperezdc** [comment]
> Yes, I think this part is ready to merge :) Is there something I need to do to "formally" request that?

We have a [request-merge-queue](https://github.com/WebKit/WebKit/labels/request-merge-queue) label in GitHub that people can add to ask that committer send a patch to the merge queue&mdash;useful for contributors who are not yet committers.

I have added this now to the queue. Thanks!

**Comment by aperezdc** [comment]
Odd, I don't know why the commit-queue bot is [trying](https://ews-build.webkit.org/#/builders/22/builds/10132/steps/20/logs/stdio) to push to the fork:

```
git push -f haiku HEAD:haiku-upstreaming
 in dir /Volumes/Data/worker/Unsafe-Merge-Queue/build (timeout 300 secs)
 watching logfiles {}
 argv: [b'git', b'push', b'-f', b'haiku', b'HEAD:haiku-upstreaming']
 using PTY: False
remote: Permission to haiku/haikuwebkit.git denied to webkit-commit-queue.
fatal: unable to access 'https://github.com/haiku/haikuwebkit.git/': The requested URL returned error: 403
program finished with exit code 128
elapsedTime=0.619070
```

**Comment by pulkomandy** [comment]
I have granted access to webkit-commit-queue following the discussion on Slack, I think it should work now?

**Comment by pulkomandy** [comment]
Since there are problems merging changes from organization owned repositories, I have resubmitted this from a user-owned fork of WebKit instead: https://github.com/WebKit/WebKit/pull/33893

**Comment by aperezdc** [comment]
Closing, this patch landed in https://github.com/WebKit/WebKit/pull/33893 in the end.

**Pull request closed without merging**